packages/clay-data-provider/README.mdx

```diff
@@ -105,6 +107,27 @@ The cache is guided by a policy, use the `fetchPolicy` prop to enable and config
 	</code> API. <b>Each new query is equivalent to 1 size</b>.
 </div>
 
+### Data Fetching
+
+```javascript
+const {resource} = useResource({fetcher, link});
+```
+
+This is an API that replaces the `link` behavior of receiving an async function, this did not allow us to validate the cache correctly because we do not have the URL view. This API is more friendly and has a unique responsibility, you may be able to pass an async function that accepts the `link` and `options`, and return the data. `useResource` will use its async function instead of the `fetch` default.
+
+#### Fetch
+
+```javascript
+import fetch from 'unfetch';
+
+const fetcher = (url, options) => fetch(url, options);
```

Comment:
This is a do-nothing pass-through wrapper, which seems like unnecessary indirection; ie. functionally equivalent to:

```js
const fetcher = fetch;
```

I would expect that if people actually wanted to use `unfetch`, they would probably write this:

```js
import fetcher from 'unfetch';

const App = () => {
  const {resource} = useResource({fetcher, link: 'https://clay.dev'});
};
```

Which begs the question, why isn't the property just called `fetch`, so that people can write?

```js
import fetch from 'unfetch';

const App = () => {
  const {resource} = useResource({fetch, link: 'https://clay.dev'});
};
```

Reply:
Yeah, this wrapper is just for a demonstration that people can add additional manipulations to their own implementation if necessary, as an example in the Storybook, this example is actually incomplete, I forgot to update. The dev must handle the return of the data, in this case:
```javascript
import unfetch from 'unfetch';

const fetcher = (url, options) => unfetch(url, options).then((res) => res.json());

const {resource} = useResource({fetcher, link: 'https://clay.dev'});
```

I think that is up to whoever is using it, in deciding to pass the fetcher without a wrapper, no problem, but he has to respect the signature:
```typescript
fetcher?: (
	link: string,
	init?: RequestInit | undefined
) => Promise<Response>;
```

Reply:
The trouble is, whatever you use as an example is what people will blindly copy-paste into the code. So, that's why I'm commenting on a useless wrapper *in documentation*, not because I am pedantic (I am), but because I don't want to see useless wrappers *in PRs* (which I'll then have to comment on there instead)...

Also, please don't forget to answer my question:

> Which begs the question, why isn't the property just called fetch

Remember that a key use case in DXP will be:

```js
import {fetch} from 'frontend-js-web';

const App = () => {
  const {resource} = useResource({fetch, link: 'https://clay.dev'});
};
```

ie. no `fetcher = fetch` wrapper; just wanting to use our `fetch`.

Related to this; it's unfortunate that the behavior changes from the old version (which called `.json()`), to the new version (which returns the response directly), because that stops the above kind of drop-in substitution from working as-is (you *would* need a wrapper, so we wind up with two wrappers: the wrapper around `fetch` so we can add our standard headers, and the wrapper just to adapt us to `useResource`).

Reply:
> Which begs the question, why isn't the property just called fetch

Oh sorry, I ended up forgetting. The word `fetcher` seems more friendly to the context in my view because we want the "fetcher" that we will use it at some point and the `fetch` sounds more that we want to get the data now, in the context of `useResource` we just want the fetcher to use it at some point and replace the standard fetcher, does that make sense to you? I think it got more complicated to understand haha ​​😅, I may be fighting with English here, in my understanding:

- `fetch` immediate action
- `fetcher` a noun, which refers to something, the function itself.

Makes sense? Fixes me if I'm wrong 😬

> The trouble is, whatever you use as an example is what people will blindly copy-paste into the code. So, that's why I'm commenting on a useless wrapper in documentation, not because I am pedantic (I am), but because I don't want to see useless wrappers in PRs (which I'll then have to comment on there instead)...

Yeah yeah, sure!! Sorry for that 🙂

> Related to this; it's unfortunate that the behavior changes from the old version (which called .json()), to the new version (which returns the response directly), because that stops the above kind of drop-in substitution from working as-is (you would need a wrapper, so we wind up with two wrappers: our existing wrapper around fetch which adds our standard headers, and the wrapper just to adapt us to useResource).

At first glance I was doing this, I went this way of internally call `.json()` for the dev, regardless of whether he is passing a custom fetch but this can get complicated if the wrapper already has some `then`, probably dev will handle this so it can do any manipulation before it reaches `useResource`, so to avoid conflicts and errors, I let dev handle it. Maybe we can validate if the return is already the given otherwise we can call `.json()` what do you think?

Reply:
> Makes sense? Fixes me if I'm wrong 😬

Yeah, I get the logic. I'm just wanting to optimize for the common case, which I *suspect* may be:

```js
useResource({fetcher: fetch, link});
```

At which point it would be easier if you could just write:

```js
useResource({fetch, link});
```

FWIW, I think JS developers are going to be pretty comfortable with functions as values, which means that they can tell the difference between `fetch` (the value) and `fetch()` (the call); they are not going to mistake the name `fetch` for the action-of-calling-`fetch`.

But this is a subtle distinction and obviously not the most important aspect of this API so I don't feel that strongly about it.

> Maybe we can validate if the return is already the given otherwise we can call `.json()` what do you think?

My main concern is that I would want the API to be as consistent and predictable as possible. The presence of the `if` in there means that we're currently doing something different based on something very subtle; eg:

```js
useResource({fetcher: window.fetch, link}) // will return the raw response
useResource({link}) // will return the JSON, even though it's also using window.fetch
```

It feels like a leaky abstraction that I have to know about the internals of how `useResource` works in order to understand that the response will be processed in two different ways depending on whether the *exact same* `fetch` implementation gets used due to being passed explicitly or used as a fallback.

Reply:
> FWIW, I think JS developers are going to be pretty comfortable with functions as values, which means that they can tell the difference between fetch (the value) and fetch() (the call); they are not going to mistake the name fetch for the action-of-calling-fetch.

Yeah, I think this is a valid argument. I'll go ahead and change the API name to `fetch`.

> My main concern is that I would want the API to be as consistent and predictable as possible. The presence of the if in there means that we're currently doing something different based on something very subtle; eg...

Yes, it makes a lot of sense. Now I'm thinking about how to make this more consistent with behavior without a custom fetcher, apparently, I wouldn't be able to circumvent the use case of having a `then` before, which can call `.json()`, so I'm thinking of ignoring it and be a limitation for that case. What do you think? I will think more about solutions before I go with this.

Writing this now, I wonder if this conditional of checking whether the data is raw or not to call `.json()` would be so bad or inconsistent 🤔.

Reply:
> Writing this now, I wonder if this conditional of checking whether the data is raw or not to call `.json()` would be so bad or inconsistent 🤔.

What would that check look like?

Reply:
What I'm thinking is:
```typescript
(res: Response) => {
	if (res.ok && res.body && !res.body.locked) {
		return res.json();
	}

	return res;
}
```

If the dev returns the raw data, we can check if the body is locked, since `res.json()` can only be called once, when called, the body is locked.

Reply:
Sorry to be so slow to reply to this one, but that seems reasonable to me, because it is consistent with the previous behavior. (On the other hand, if the old version *didn't* call `json()` and we started calling it, then that would feel like a breaking change.)

Reply:
No problem. Yeah, sure. I'll get on with that.